tokio-net/src/driver/sharded_slab/mod.rs

```diff
@@ -0,0 +1,317 @@
+            idx += 1;
+            CausalCell::new(shard)
+        });
+        Slab {
+            shards: shards.into_boxed_slice(),
+        }
+    }
+
+    /// Inserts a value into the slab, returning a key that can be used to
+    /// access it.
+    ///
+    /// If this function returns `None`, then the shard for the current thread
+    /// is full and no items can be added until some are removed, or the maximum
+    /// number of shards has been reached.
+    pub(crate) fn insert(&self, value: T) -> Option<usize> {
+        let tid = Tid::current();
+        self.shards[tid.as_usize()]
+            .with_mut(|shard| unsafe {
+                // we are guaranteed to only mutate the shard while on its thread.
+                (*shard).insert(value)
+            })
+            .map(|idx| tid.pack(idx))
+    }
+
+    /// Removes the value associated with the given key from the slab, returning
+    /// it.
+    ///
+    /// If the slab does not contain a value for that key, `None` is returned
+    /// instead.
+    pub(crate) fn remove(&self, idx: usize) -> Option<T> {
+        let tid = Tid::from_packed(idx);
+        if tid.is_current() {
+            self.shards[tid.as_usize()].with_mut(|shard| unsafe {
+                // only called if this is the current shard
+                (*shard).remove_local(idx)
+            })
+        } else {
+            self.shards[tid.as_usize()].with(|shard| unsafe { (*shard).remove_remote(idx) })
+        }
+    }
+
+    /// Return a reference to the value associated with the given key.
+    ///
+    /// If the slab does not contain a value for the given key, `None` is
+    /// returned instead.
+    pub(crate) fn get(&self, key: usize) -> Option<&T> {
+        let tid = Tid::from_packed(key);
+        self.shards
+            .get(tid.as_usize())?
+            .with(|shard| unsafe { (*shard).get(key) })
+    }
+
+    /// Returns the number of items currently stored in the slab.
+    pub(crate) fn len(&self) -> usize {
+        self.shards
+            .iter()
+            .map(|shard| shard.with(|shard| unsafe { (*shard).len() }))
+            .sum()
+    }
+
+    /// Returns an iterator over all the items in the slab.
+    pub(crate) fn unique_iter(&mut self) -> iter::UniqueIter<'_, T> {
+        let mut shards = self.shards.iter_mut();
+        let shard = shards.next().expect("must be at least 1 shard");
+        let mut pages = shard.with(|shard| unsafe { (*shard).iter() });
+        let slots = pages.next().expect("must be at least 1 page").iter();
+        iter::UniqueIter {
+            shards,
+            slots,
+            pages,
+        }
+    }
+}
+
+impl<T> Shard<T> {
+    fn new(_tid: usize) -> Self {
+        Self {
+            #[cfg(debug_assertions)]
+            tid: _tid,
+            sz: page::INITIAL_SIZE,
+            len: AtomicUsize::new(0),
+            pages: vec![Page::new(page::INITIAL_SIZE, 0)],
+        }
+    }
+
+    fn insert(&mut self, value: T) -> Option<usize> {
```

Comment:
I think technically taking `&mut self` in these functions is undefined because `remove_remote` can be called concurrently. At any point a `&mut` reference comes into existence, it must be guaranteed that it is a unique reference (not aliased).

Usually I solve this by splitting structs into an `Owned` and `Shared` (or something like that). i.e. data that can only be accessed by a single thread and data that can be accessed concurrently. Additionally, this is often a good strategy to avoid false sharing. It **looks** like your `Page` structure is at risk for false sharing.

I'd consider a `Shard` struct like:

```rust
struct Shard<T> {
   // ... stuff
   page_local: Vec<page::Local>,
   page_shared: Vec<page::Shared<T>>,
}

mod page {
    struct Local {
        // fields that are only used from main thread and need `&mut self`.
    }

    struct Shared<T> {
        remote_head: AtomicUsize,
        slots: Box<[Slot<T>]>,
   }
}
```

Reply:
Thanks, this is a really good idea!

The free lists used to have a tail as well as a head (the tail was removed as it was never used), so they were effectively padded to a cache line; I didn't notice that removing the tails would put the local and remote heads in the same cache line.

Reply:
You can't really assume correct alignment without an attr annotation. Otherwise, the compiler is free to do whatever.

Something like

```rust
#[cfg_attr(target_arch = "x86_64", repr(align(128)))]
#[cfg_attr(not(target_arch = "x86_64"), repr(align(64)))]
```

Reply:
Yeah. Regardless, I think your suggestion of just putting them in separate Vecs is probably best, since it also avoids potential UB.